keps/sig-api-machinery/4988-serve-pagination-from-cache/README.md

```diff
@@ -0,0 +1,401 @@
+to fetch the next page.
+
+To serve paginated requests correctly, the API server needs access to a consistent
+snapshot of the data from which the initial page was generated. Watchcache does
+not store historical data. Therefore, currently, the API server does the following:
+
+- For `resourceVersion=""`: Serve paginated requests directly from etcd, bypassing the cache.
+- For `resourceVersion="0"`: Ignore the limit parameter and return the entire list in a single response.
+
+Both options are problematic:
+
+- Serving from etcd negates the benefits of the cache,
+  leading to increased load on both the API server and etcd.
+- Ignoring the limit parameter defeats the purpose of pagination,
+  potentially returning massive responses that can overwhelm clients.
+
+This situation leads to unexpected behavior for users,
+where enabling pagination can degrade performance instead of improving it.
+
+### Goals
+
+- Improve the performance of paginated list requests to be comparable to non-paginated lists.
+- Make pagination behavior predictable and consistent.
+- Reduce load on etcd caused by paginated requests.
+
+### Non-Goals
+
+- Serve `resourceVersion="N"` request from watch cache
+- Support indices when paginating.
+- Eliminate all paginated list requests to etcd.
+
+## Proposal
+
+Leveraging the recent rewrite of the watchcache storage layer to use a B-tree
+(https://github.com/kubernetes/kubernetes/pull/126754), we propose to utilize
+B-tree snapshots to serve paginated lists.
+
+Mechanism:
+1. **Snapshot Creation:** When a paginated list request (with a limit parameter
+   set) is received, the API server will create a snapshot of the B-tree based cache
+   using the efficient [Clone()](https://pkg.go.dev/github.com/google/btree#BTree.Clone) method.
+   This clone is a lazy copy, only duplicating necessary nodes, resulting in
+   minimal overhead and effectively is container of pointers in-memory and not actual copies
+   of the B-tree nodes. Since we do not edit event histories in the watchCache, a snapshot
+   is treated as a read-only entity and its creation is cheap (creating only pointers in memory).
+2. **Snapshot Storage:** The snapshot will be stored in memory, keyed by
+   resourceVersion the watchCache has propagated upto when the snapshot was created. It is also this
+   resourceVersion that is returned as part of the continuation token in the response. 
+3. **Serving Subsequent Pages:** When a subsequent request with a continue token
+   arrives, the API server will:
+  - Extract the resourceVersion from the continue token. 
+  - Since resourceVersions provide a global logical clock sequencing all events in the cluster, a snapshot
+    of the watchCache for this resourceVersion is retrieved using the resourceVersion as the key.
+  - The corresponding snapshot may not be present in the following 2 scenarios at an API Server:
+    - Snapshot has been cleaned up due to the 75s TTL (see below).
```

Comment:
nit: we switched that recently, with 75s still being default, but it now depends on request timeouts

Reply:
Right